src/main/scala/chisel3/util/Inline.scala

```diff
@@ -0,0 +1,96 @@
+  *  class Bar extends Module with Internals
+  *  class Top extends Module {
+  *    val io = IO(new Bundle{ val a = Input(Bool()) })
+  *    val x = Module(new Foo)                     // x will be inlined
+  *    val y = Module(new Bar with InlineInstance) // y will also be inlined
+  *    val z = Module(new Bar)                     // z will not be inlined
+  *    Seq(x, y, z).map(_.io.a := io.a)
+  *  }
+  * }}}
+  */
+trait InlineInstance { self: BaseModule =>
+  chisel3.experimental.annotate(
+    new ChiselAnnotation with RunFirrtlTransform {
+      def toFirrtl = InlineAnnotation(self.toNamed)
+      def transformClass = classOf[InlineInstances] } )
+  chisel3.experimental.annotate(
+    new ChiselAnnotation {
+      def toFirrtl = NoDedupAnnotation(self.toNamed) } )
+}
+
+/** Provide methods for controlling inlining from within a [[chisel3.experimental.BaseModule]]
+  */
+trait InlineHelpers { self: BaseModule =>
+  /* If used anywhere, this will result in every module in the design being
+   * inlined. The resulting circuit will contain only one module. */
+
+  /** This, if used anywhere, will result in the entire circuit being
+    * flattened to a single module.
+    *
+    * @example {{{
+    * /* Providing some module internals */
+    * trait Internals { this: Module =>
+    *   val io = IO(new Bundle{ val a = Input(Bool()) })
+    * }
+    * class Bar extends Module with Internals with InlineHelpers
+    * class Top extends Module with InlineHelpers {
+    *   val io = IO(new Bundle{ val a = Input(Bool()) })
+    *   val Seq(x, y, z) = Seq.fill(3)(Module(new Bar))
+    *   Seq(x, y, z).map(_.io.a := io.a)
+    *   /* All modules in the entire circuit will be inlined. Here, that is x, y,
+    *    * and z. */
+    *   inlineCircuit()
+    * }
+    * }}}
+    */
+  def inlineCircuit(): Unit = chisel3.experimental.annotate(
+    new ChiselAnnotation with RunFirrtlTransform {
+      def toFirrtl = InlineAnnotation(self.toNamed match { case ModuleName(_, c) => c } )
+      def transformClass = classOf[InlineInstances] } )
+
+  /** Inline a specific instance inside a Module
+    *
+    * @param m the module to inline
+    * @example {{{
+    * /* Providing some module internals */
+    * trait Internals { this: Module =>
+    *   val io = IO(new Bundle{ val a = Input(Bool()) })
+    * }
+    * class Bar extends Module with Internals with InlineHelpers
+    * class Top extends Module with InlineHelpers {
+    *   val io = IO(new Bundle{ val a = Input(Bool()) })
+    *   val Seq(x, y, z) = Seq.fill(3)(Module(new Bar))
+    *   Seq(x, y, z).map(_.io.a := io.a)
+    *   /* Only the insance x will be inlined */
+    *   inlineComponent(x)
+    * }
+    * }}}
+    */
+  def inlineComponent(m: BaseModule): Unit = chisel3.experimental.annotate(
+    new ChiselAnnotation with RunFirrtlTransform {
+      def toFirrtl = InlineAnnotation(ComponentName(m.instanceName, self.toNamed))
+      def transformClass = classOf[InlineInstances] }
```

Comment:
What is the utility of this alternative way of specifying that an instance should be inlined? I guess you could make a submodule in code you don't control or don't want to modify inline an instance.

Reply:
Had to think about this as I didn't have good justification originally. If a library writer chooses to only expose a wrapped `Module` via a factory then this would be the only way to inline that instance. Consider:
```scala
object BarFactory {
  def apply(): Bar = Module(new Bar)
}
class Top extends Module with InlineHelpers {
  val io = IO(new Bundle{ val a = Input(Bool()) })
  val Seq(x, y, z) = Seq.fill(3)(BarFactory())
  Seq(x, y, z).map(_.io.a := io.a)
  inlineComponent(x)
}
```
There's nothing that I'm aware of where this type of API is the only one that is exposed (something like `chisel3.util.Queue` is generally used only via factory, but the underlying class is public). Nonetheless, there may be some motivation for only exposing a factory and causing that.

Do you think this is worth including or is the case here too pathological?

Reply:
As a user I think it would be good to be able to inline a module created in a 3rd-party library without having to modify that library (promotes re-use instead of fork/copy-paste).

Reply:
Sorry I missed this 10 days ago.

I agree that we would like to be able to inline modules constructed in factories methods, but in practice this won't work for that either. As in the case of `Queue`, factory methods for constructing modules typically don't return the actual module reference itself, rather they return an interface (eg. see [object Queue](https://github.com/freechipsproject/chisel3/blob/98177d3647e51efe869de1031872adb6406fa34d/src/main/scala/chisel3/util/Decoupled.scala#L276)).

I think we need to come up with another way to support that in the general case. One idea I (or @aswaterman, I can't remember who proposed it) had was to provide a mechanism similar to `withClock` where you could scope things below to be inlined. There could be something like `inlineInstances { }` or the like. Unfortunately, this would require some form of support from Chisel, either natively or some API to get references to the Modules instantiated within a block.

In any case, I think we should remove this API and just have the mixin.

Reply:
Yeah, `Queue` was only intending to be illustrative of how a user might write a library (and hopefully pass back the module as opposed to the IO). I like the idea of `inlineInstances { }`, but yes, that's likely out of scope here.

As API deprecation is such a hot topic, I think it's best to err with new APIs on the side of caution. I'll remove this.